pkgs/tools/networking/haproxy/default.nix

```diff
@@ -1,22 +1,39 @@
 { useLua ? true
 , usePcre ? true
 , withPrometheusExporter ? true
+, sslLibrary ? "wolfssl"
 , stdenv
 , lib
 , fetchurl
 , nixosTests
-, openssl
 , zlib
 , libxcrypt
-, lua5_3 ? null
-, pcre ? null
+, wolfssl-haproxy ? null
+, libressl ? null
+, quictls ? null
+, boringssl ? null
+, openssl ? null
+, lua5_4 ? null
+, pcre2 ? null
 , systemd ? null
 }:
 
-assert useLua -> lua5_3 != null;
-assert usePcre -> pcre != null;
+assert useLua -> lua5_4 != null;
+assert usePcre -> pcre2 != null;
+assert lib.assertOneOf "sslLibrary" sslLibrary [ "wolfssl" "libressl" "quictls" "boringssl" "openssl" ];
+assert sslLibrary == "wolfssl" -> wolfssl-haproxy != null;
+assert sslLibrary == "libressl" -> libressl != null;
+assert sslLibrary == "quictls" -> quictls != null;
+assert sslLibrary == "boringssl" -> boringssl != null;
+assert sslLibrary == "openssl" -> openssl != null;
 
-stdenv.mkDerivation (finalAttrs: {
+let
+  sslPkgs = {
+    inherit libressl quictls boringssl openssl;
+    wolfssl = wolfssl-haproxy;
+  };
+  sslPkg = sslPkgs.${sslLibrary};
```

Comment:
We should probably just override the packages and override openssl like we do for nginx than this other layer of abstraction.